File: lab10/j.py
import sys
from collections import deque 
sys.setrecursionlimit(10 ** 7) 
input = sys.stdin.readline 



n , m = map(int , input().split())
g = [[] for _ in range(n+1)]
for _ in range(m):
    x , y = map(int , input().split())
    g[x].append(y)
    g[y].append(x) 

visited = [False] * (n+1)
bigFam = 0 

def dfs_component(root):
    comp = []
    stack = [root]
    visited[root] = True 
    while stack:
        u = stack.pop() 
        comp.append(u)
        for v in g[u]:
            if not visited[v]:
                visited[v] = True 
                stack.append(v)
    return comp  

def build_tree_and_count(root , comp):
    parent = {root: -1}
    chiildren_count = {v:0 for v in comp}

    q = deque([root])
    visited2 = set([root])

    while q:
        u = q.popleft()
        for v in g[u]: 
            if v not in visited2:
                visited2.add(v)
                parent[v] = u
                chiildren_count[u] += 1 
                q.append(v) 

    return parent , chiildren_count 

answer = 0 

for i in range(1 , n+1):
    if not visited[i]:
        comp = dfs_component(i)

        root = min(comp)

        parent , children = build_tree_and_count(root , comp)

        for v in comp:
            if v == root:
                answer +=1  
            else:
                p = parent[v]
                if children[v] > children[p]:
                    answer += 1 
print(answer)

File: lab10/j.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m;
    cin >> n >> m;

    vector<vector<int>> g(n+1);
    for (int i = 0; i < m; i++) {
        int x, y;
        cin >> x >> y;
        g[x].push_back(y);
        g[y].push_back(x);
    }

    vector<bool> visited(n+1, false);
    int answer = 0;

    function<vector<int>(int)> get_component = [&](int start) {
        vector<int> comp;
        stack<int> st;
        st.push(start);
        visited[start] = true;

        while (!st.empty()) {
            int u = st.top(); st.pop();
            comp.push_back(u);
            for (int v : g[u]) {
                if (!visited[v]) {
                    visited[v] = true;
                    st.push(v);
                }
            }
        }
        return comp;
    };

    for (int i = 1; i <= n; i++) {
        if (!visited[i]) {
            vector<int> comp = get_component(i);

            int root = *min_element(comp.begin(), comp.end());

            // BFS to make the tree
            unordered_map<int,int> parent;
            unordered_map<int,int> children;
            for (int v : comp) children[v] = 0;

            queue<int> q;
            unordered_set<int> seen;
            q.push(root);
            seen.insert(root);
            parent[root] = -1;

            while (!q.empty()) {
                int u = q.front(); q.pop();
                for (int v : g[u]) {
                    if (!seen.count(v)) {
                        seen.insert(v);
                        parent[v] = u;
                        children[u]++;
                        q.push(v);
                    }
                }
            }

            // Count BigFam
            for (int v : comp) {
                if (v == root) answer++;
                else {
                    int p = parent[v];
                    if (children[v] > children[p])
                        answer++;
                }
            }
        }
    }

    cout << answer << "\n";
    return 0;
}
